### PR TITLE
[Serve] Use Flask request class for incoming request

### PR DESCRIPTION
**Pull request by simon-mo**
Please leave feedback for this direction:
- I believe flask.Request class is well-structured and well-understood enough to be directly adopted. 
- Flask users can quickly look up what they wanted from the flask doc. 
- Lower the barrier to entry.

### CODE REVIEW

File: python/ray/experimental/serve/server.py [hunk 1]
```diff
@@ -1,14 +1,67 @@
...
+        environ["SERVER_NAME"] = scope["server"][0]
+        environ["SERVER_PORT"] = str(scope["server"][1])
+    else:
+        environ["SERVER_NAME"] = "localhost"
+        environ["SERVER_PORT"] = "80"
+
+    if "client" in scope:
+        environ["REMOTE_ADDR"] = scope["client"][0]
+
+    # Go through headers and make them into environ entries
```
Comment (edoakes) [inline python/ray/experimental/serve/server.py]:
```suggestion
    # Transforms headers into environ entries.
```

File: python/ray/experimental/serve/server.py [hunk 2]
```diff
@@ -1,14 +1,67 @@
...
+    for name, value in scope.get("headers", []):
+        name = name.decode("latin1")
+        if name == "content-length":
+            corrected_name = "CONTENT_LENGTH"
+        elif name == "content-type":
+            corrected_name = "CONTENT_TYPE"
+        else:
+            corrected_name = "HTTP_%s" % name.upper().replace("-", "_")
+        # HTTPbis say only ASCII chars are allowed in headers, but we latin1 just in case
```
Comment (edoakes) [inline python/ray/experimental/serve/server.py]: What does "just in case mean" ? Shouldn't we just reject anything that's non-ascii?

File: python/ray/experimental/serve/server.py [hunk 3]
```diff
@@ -29,6 +82,8 @@ def __init__(self, content=None, status_code=200):
         self.status_code = status_code
         self.raw_headers = [[b"content-type", b"application/json"]]
 
+        ray.register_custom_serializer(flask.Request, use_pickle=True)
```
Comment (edoakes) [inline python/ray/experimental/serve/server.py]: why is this necessary? comment please

File: python/ray/experimental/serve/server.py [hunk 4]
```diff
@@ -1,14 +1,67 @@
...
+        environ["SERVER_PORT"] = "80"
+
+    if "client" in scope:
+        environ["REMOTE_ADDR"] = scope["client"][0]
+
+    # Go through headers and make them into environ entries
+    for name, value in scope.get("headers", []):
+        name = name.decode("latin1")
+        if name == "content-length":
+            corrected_name = "CONTENT_LENGTH"
```
Comment (edoakes) [inline python/ray/experimental/serve/server.py]: Please link to the documentation that specifies these transformations

File: python/ray/experimental/serve/utils.py [hunk 5]
```diff
@@ -29,6 +30,11 @@ class BytesEncoder(json.JSONEncoder):
     def default(self, o):  # pylint: disable=E0202
         if isinstance(o, bytes):
             return o.decode("utf-8")
+        import flask
+        if isinstance(o, flask.Request):
+            return o.__dict__
+        if isinstance(o, io.BytesIO):
```
Comment (edoakes) [inline python/ray/experimental/serve/utils.py]: Why do we need to handle both `bytes` and `io.BytesIO`?

**Review by edoakes — COMMENTED** [review]
Good idea! Looks pretty good, please re-request once you've addressed.

**Comment by simon-mo** [comment]
Closed in favor of a new upcoming PR.
